app/src/main/java/org/openobservatory/ooniprobe/activity/runtests/adapter/RunTestsExpandableListViewAdapter.kt

```diff
@@ -0,0 +1,229 @@
+					childItem.selected = true
+				}
+				if (isSelectedAllItems(mGroupListData)) {
+					mViewModel.setSelectedAllBtnStatus(SELECT_ALL)
+				} else {
+					mViewModel.setSelectedAllBtnStatus(SELECT_SOME)
+				}
+			}
+			notifyDataSetChanged()
+		}
+		if (isExpanded) {
+			groupIndicator.setImageResource(R.drawable.expand_less)
+		} else {
+			groupIndicator.setImageResource(R.drawable.expand_more)
+		}
+		return convertView
+	}
+
+	override fun getChildView(
+		groupPosition: Int,
+		childPosition: Int,
+		isLastChild: Boolean,
+		convertView: View?,
+		parent: ViewGroup
+	): View? {
+		var convertView =
+			convertView ?: LayoutInflater.from(mContext).inflate(R.layout.run_tests_child_list_item, parent, false)
+		val childItem = getChild(groupPosition, childPosition)
+		val groupItem = getGroup(groupPosition)
+		val nettest = AbstractTest.getTestByName(childItem.name)
+		convertView.findViewById<TextView>(R.id.child_name)?.apply {
+			text = when (groupItem.name) {
+				ExperimentalSuite.NAME -> {
+					childItem.name
+				}
+
+				else -> {
+					parent.context.resources.getText(nettest.labelResId)
+				}
+			}
+		}
+		convertView.findViewById<ImageView>(R.id.child_select).apply {
+			setImageResource(
+				when (childItem.selected) {
+					true -> R.drawable.check_box
+					false -> R.drawable.check_box_outline_blank
+				}
+			)
+			setOnClickListener {
+				if (childItem.selected) {
+					childItem.selected = false
+					mViewModel.disableTest(childItem.name)
+					if (isNotSelectedAnyChildItems(groupItem.nettests)) {
+						groupItem.selected = false
+					}
+					if (isNotSelectedAnyItems(mGroupListData)) {
+						mViewModel.setSelectedAllBtnStatus(SELECT_NONE)
+					} else {
+						mViewModel.setSelectedAllBtnStatus(SELECT_SOME)
+					}
+				} else {
+					childItem.selected = true
+					mViewModel.enableTest(childItem.name)
+					groupItem.selected = true
+					if (isSelectedAllItems(mGroupListData)) {
+						mViewModel.setSelectedAllBtnStatus(SELECT_ALL)
+					} else {
+						mViewModel.setSelectedAllBtnStatus(SELECT_SOME)
+					}
+				}
+				notifyDataSetChanged()
+			}
+		}
+		return convertView
+	}
+
+	override fun isChildSelectable(groupPosition: Int, childPosition: Int): Boolean = false
+
+	private fun isNotSelectedAnyGroupItem(groupItemsList: List<GroupItem>): Boolean {
+		for (groupItem in groupItemsList) {
+			if (groupItem.selected) {
+				return false
+			}
+		}
+		return true
+	}
+
+	private fun isNotSelectedAnyChildItems(childItemList: List<ChildItem>): Boolean {
+		for (childItem in childItemList) {
+			if (childItem.selected) {
+				return false
+			}
+		}
+		return true
+	}
+
+	private fun isSelectAllChildItems(childItemList: List<ChildItem>): Boolean {
+		for (childItem in childItemList) {
+			if (!childItem.selected) {
+				return false
+			}
+		}
+		return true
+	}
+
+	private fun isSelectedAllItems(groupItemList: List<GroupItem>?): Boolean {
+		for (groupItem in groupItemList!!) {
```

Comment:
What is the meaning of the trailing `!!`?

Reply:
This is used to verify that the `groupItemList` is not null

Reply:
Why do we declare `List<GroupItem>?` if then we `!!` which throws if `null`? Wouldn't it be better to declare `List<GroupItem>` directly?

Reply:
technically it can't be null but I will adjust the code to address the issue